src/runtimes/index.js

```diff
@@ -1,8 +1,82 @@
+const { extname, basename } = require('path')
+
 const goRuntime = require('./go')
 const jsRuntime = require('./node')
 const rustRuntime = require('./rust')
 
+/**
+ * Finds functions for a list of paths using a specific runtime. The return
+ * value is an object containing an array of the functions found (`functions`)
+ * and an array with the paths that haven't been recognized by the runtime
+ * (`remainingPaths`).
+ *
+ * @param   {Array<String>} paths
+ * @param   {Object} runtime
+ * @returns {Promise<Object>}
+ */
+const findFunctionsInRuntime = async function ({ dedupe = false, paths, runtime }) {
+  const functions = await runtime.findFunctionsInPaths(paths)
+
+  // If `dedupe` is true, we use the function name (`filename`) as the map key,
+  // so that `function-1.js` will overwrite `function-1.go`. Otherwise, we use
+  // `srcPath`, so that both functions are returned.
+  const key = dedupe ? 'name' : 'srcPath'
+
+  // Augmenting the function objects with additional information.
+  const augmentedFunctions = functions.map((func) => [
+    func[key],
+    {
+      ...func,
+      extension: extname(func.mainFile),
+      filename: basename(func.srcPath),
+      runtime,
+    },
+  ])
+  const usedPaths = new Set(augmentedFunctions.map(([path]) => path))
+  const remainingPaths = paths.filter((path) => !usedPaths.has(path))
+
+  return { functions: augmentedFunctions, remainingPaths }
+}
+
+/**
+ * Gets a list of functions found in a list of paths.
+ *
+ * @param   {Boolean} dedupe
+ * @param   {String} path
+ * @returns {Promise<Map>}
+ */
+const getFunctionsFromPaths = async (paths, { dedupe = false } = {}) => {
+  // The order of this array determines the priority of the runtimes. If a path
+  // is used by the first time, it won't be made available to the subsequent
+  // runtimes.
+  const runtimes = [jsRuntime, goRuntime, rustRuntime]
+
+  // We cycle through the ordered array of runtimes, passing each one of them
+  // through `findFunctionsInRuntime`. For each iteration, we collect all the
+  // functions found plus the list of paths that still need to be evaluated,
+  // using them as the input for the next iteration until the last runtime.
+  const { functions } = await runtimes.reduce(
+    async (aggregate, runtime) => {
+      const { functions: aggregateFunctions, remainingPaths: aggregatePaths } = await aggregate
+      const { functions: runtimeFunctions, remainingPaths: runtimePaths } = await findFunctionsInRuntime({
+        dedupe,
+        paths: aggregatePaths,
+        runtime,
+      })
+
+      return {
+        functions: [...aggregateFunctions, ...runtimeFunctions],
+        remainingPaths: runtimePaths,
+      }
+    },
+    { functions: [], remainingPaths: paths },
+  )
+
+  return new Map(functions)
```

Comment:
Every consumer of this function is iterating over it with `map.values()`.
Would it make sense to directly return `[...new Map(functions).values()]`?

Reply:
I'm a bit torn on this, actually.

On the one hand, it would simplify the consumers.

On the other hand, we'd be "losing information" by converting it to an array prematurely — e.g. we lose the ability to query a specific function by name, if we encounter that use case. Also, converting the Map to an array comes with a performance price which we don't always have to pay. For example, [here](https://github.com/netlify/zip-it-and-ship-it/blob/feat/typescript-support/src/zip.js#L86) we can simply access the right element in the Map by making use of the iterator, without having to convert to an array.

I don't have any strong preferences though, so I'm happy to make the change if you do!

Reply:
This makes sense, good point :+1: